docs/proposals/agent-pkg-tagset-rfc.md

```diff
@@ -0,0 +1,303 @@
+
+#### Factory
+
+The Factory type is responsible for creating new Tags instances.
+Its interface is simple, but provides an opportunity for optimization and deduplication.
+
+As with many Go packages, a single default factory is provided for use throughout the agent, with package-level functions deferring to that factory.
+Additional, specific factories may be created for specific purposes.
+The default factory is thread-safe, but it may be advantageous to build non-thread-safe factories for specific circumstances.
+Tags instances created by different factories can be used interchangeably and are entirely thread-safe.
+
+ * Constructors
+   * `Factory.NewTags(src []string) *Tags` - create a new Tags, leaving ownership of the slice with the caller (so modifying the slice after calling this method will not modify the resulting Tags instance)
+   * `Factory.NewTagsFromMap(src map[string]struct{}) *Tags` - create a new Tags based on the keys of the given map (also implying uniqueness)
+   * `Factory.NewTag(tag string) *Tags` - create a new Tags, containing only the given tag
+   * `Factory.NewBuilder(capacity int) Builder` - create a new Builder, tied back to this factory for caching purposes
+   * `Factory.NewSliceBuilder(levels, capacity int) SliceBuilder` - create a new SliceBuilder, tied back to this factory for caching purposes
+ * Parsing
+   * `Factory.UnmarshalJSON(data []byte) (Tags, error)`
+   * `Factory.UnmarshalYAML(data []byte) (Tags, error)`
+   * `Factory.ParseDSD(data []byte) (Tags, error)` - Parse tags in DogStatsD format (comma-separated)
+ * Combination
+   * `Factory.Union(a, b Tags) *Tags` - perform a union of two Tags instances, which may contain the same tags
+   * `Factory.DisjointUnion(a, b Tags) *Tags` - like Union, but with the caller promising that the sets are disjoint
+
+#### Builders
+
+Builders are used to build tagsets tag-by-tag, before "freezing" into one or more `Tags` instances.
+Builders are not threadsafe.
+
+A builder goes through three stages in its lifecycle: 
+ 1. adding tags (begins on `factory.New*Builder`).
+ 2. frozen (begins on call to `acc.Freeze*`).
+ 3. closed (begins on call to `acc.Close`).
+
+The `Add` methods may only be called in the first stage.
+No methods may be called in the third stage.
+
+The Builder type supports the common use-case of conditionally adding a handful of tags to a set.
+Adding the same tag twice to an Builder is a no-op.
+
+The SliceBuilder type supports creation of overlapping "slices" of tags.
+This supports Tagger's tag-cardinality functionality while saving some memory space.
+For example, a SliceBuilder can produce a Tags containing high, orchestrator, and low-cardinality tags as well as a Tags with only the low-cardinality tags, sharing that space.
+Adding the same tag twice to a SliceBuilder is a no-op if both are at the same level, but results in undefined behavior if the levels differ.
```

Comment:
Let's not have undefined behavior :) For one, UB is a scary phrase if taken in the C sense, and even if not, it isn't very helpful.

I am also curious about this level concept. What does it do? It doesn't seem to be defined in the RFC as far as I can tell. How will it be used?

Reply:
SliceBuilder is described [here in the RFC](https://github.com/DataDog/architecture/blob/master/rfcs/agent-pkg-tagset/rfc.md#builders).

This is used by the tagger, which is currently not very perf-sensitive, so I can add some checks for adding the same tag twice (it just means keeping some extra maps and checking/inserting into them).

Reply:
> I am also curious about this level concept. What does it do? It doesn't seem to be defined in the RFC as far as I can tell. How will it be used?

+1 I am still reading the code but I am not finding a clear definition of what a `level` and how it's intended to be used.

Reply:
Can you indicate what's missing from my previous comment?

Reply:
Here's an example of how it's used: https://github.com/DataDog/datadog-agent/blob/23cb0c5fd9b3bc1b420ca7e810786c52926662b5/pkg/tagger/tagstore/entity_tags.go#L124-L138